lib/Driver/ToolChains.cpp

```diff
@@ -847,6 +848,74 @@ toolchains::Darwin::constructInvocation(const InterpretJobAction &job,
   return II;
 }
 
+static StringRef
+getDarwinLibraryNameSuffixForTriple(const llvm::Triple &triple) {
+  switch(getDarwinPlatformKind(triple)) {
+    case DarwinPlatformKind::MacOS:
+      return "osx";
+    case DarwinPlatformKind::IPhoneOS:
+      return "ios";
+    case DarwinPlatformKind::IPhoneOSSimulator:
+      return "iossim";
+    case DarwinPlatformKind::TvOS:
+      return "tvos";
+    case DarwinPlatformKind::TvOSSimulator:
+      return "tvossim";
+    case DarwinPlatformKind::WatchOS:
+      return "watchos";
+    case DarwinPlatformKind::WatchOSSimulator:
+      return "watchossim";
+  }
+  llvm_unreachable("Unsupported Darwin platform");
+}
+
+static void AddLinkRuntimeLib(const ArgList &Args, ArgStringList &Arguments,
+                              StringRef DarwinLibName, bool AddRPath,
+                              const ToolChain &TC) {
+  SmallString<128> Dir;
+  getRuntimeLibraryPath(Dir, Args, TC);
+  // Remove platform name.
+  llvm::sys::path::remove_filename(Dir);
+  llvm::sys::path::append(Dir, "clang", "lib", "darwin");
+  SmallString<128> P(Dir);
+  llvm::sys::path::append(P, DarwinLibName);
+  Arguments.push_back(Args.MakeArgString(P));
+
+  // Adding the rpaths might negatively interact when other rpaths are involved,
+  // so we should make sure we add the rpaths last, after all user-specified
+  // rpaths. This is currently true from this place, but we need to be
+  // careful if this function is ever called before user's rpaths are emitted.
+  if (AddRPath) {
+    assert(DarwinLibName.endswith(".dylib") && "must be a dynamic library");
+
+    // Add @executable_path to rpath to support having the dylib copied with
+    // the executable.
+    Arguments.push_back("-rpath");
+    Arguments.push_back("@executable_path");
+
+    // Add the path to the resource dir to rpath to support using the dylib
+    // from the default location without copying.
+    Arguments.push_back("-rpath");
+    Arguments.push_back(Args.MakeArgString(Dir));
+  }
+}
+
+static void AddLinkSanitizerLibArgs(const ArgList &Args,
```

Comment:
Since this implementation will not work on non-Darwin, please put "Darwin" in the name. (We don't know which C++ stdlib and ABI libraries to use elsewhere, for one.)